### DISCUSSION TITLE
Problem serializing Postgres' array columns due to delegation.

### DISCUSSION THREAD

**Discussion by tomasmiguez** [jeremyevans/sequel#2153] [General]
Hi!
We are currently working with Sequel as our ORM, Postgres as our DBMS and [Oj](https://github.com/ohler55/oj) for dumping objects to JSON. This works fine on most use cases but we recently started using array type columns and started experiencing some issues.

For example, if we have this table
```
                Table "public.tests"
 Column |   Type    | Collation | Nullable | Default
--------+-----------+-----------+----------+---------
 foo    | integer[] |           |          |
```

and the model (the `pg_array` extension is already loaded)

```ruby
require 'sequel'
  
class Test < Sequel::Model
end
```

when accesing the column we get:

```ruby
irb:004> Test.first.foo
=> [1, 2, 3, 4, 5]
```

but that's not an array, it's a `PGArray` that delegates to `Array`. This causes a problem when dumping with Oj, as that gem serializes the instance variables of the object that  is being dumped, and we get

```ruby
irb:006> puts Oj.dump(Test.first.foo)
{"^o":"Sequel::Postgres::PGArray","delegate_dc_obj":[1,2,3,4,5],"array_type":":integer"}
=> nil
```

which is not the expected result. We checked the [source code of the extension](https://github.com/jeremyevans/sequel/blob/master/lib/sequel/extensions/pg_array.rb) from where we got the idea to try to solve this issue by registering a conversion proc with `add_conversion_proc`, but without success. We also tried different option on the dumping of Oj, mainly `use_to_hash` and `use_to_json` but this didn't help.

The current solution we are using is to explicitly get the delegate object of the column on the accesor

```ruby
  class Test < Sequel::Model
    def foo
      super.__getobj__
    end
  end
```

which works, but we would rather not have to override that for every array column of every model. 

Is there any mechanism in Sequel that might help us solve this issue? Maybe a way to wrap the accesors methods of every column of every model for a specific type or something like that?

Thanks!

**Comment by jeremyevans** [comment]
I don't think something is built into Sequel that will do what you want.  Having the access method return the underlying array instead of the wrapper could definitely break things in certain cases (though maybe not cases that would affect you).

I've never used Oj, so I don't know how much control it gives over the dumping process.  It would be best to fix the issue in Oj, since the problem is Oj not handling delegate objects as you want (this isn't a Sequel-specific issue, it likely affects any use of the delegate classes using the stdlib delegate library). Does setting the `create_id: nil`  option when calling `Oj.dump` fix things?

**Reply by tomasmiguez** [reply]
Thanks Jeremy! We will try that right away or check if we can fix it from Oj's side.

**Comment by rgalanakis** [comment]
I know this doesn't directly answer your question, but we use PG array columns and `Oj` in [webhookdb](https://webhookdb.com). It took some fiddling to get the JSON dumping right (all the knobs are sort of confusing), here is the code we ended up with, in case it's useful: https://github.com/webhookdb/webhookdb/blob/main/lib/webhookdb/json.rb

**Reply by tomasmiguez** [reply]
Thanks for the info! We would like to keep custom logic for this to a minimum and use Oj's logic as much as we can for performance reasons, but we will keep this in mind as an option.
